mut/redirects/redirect_main.py

```diff
@@ -0,0 +1,396 @@
+"""
+Usage:
+    mut-redirects <source_path> -o <output>
+
+    -h, --help             List CLI prototype, arguments, and options.
+    <source_path>          Path to the file(s) containing redirect rules.
+    -o, --output <output>  File path for the output .htaccess file.
+"""
+
+# Spec URL: https://docs.google.com/document/d/1oI2boFmtzvbbvt-uQawY9k_gLSLbW7LQO2RjVkvtRgg/edit?ts=57caf48b#
+
+import re
+from docopt import docopt
+
+rules = []
+symlinks = []
+definitions = []
+
+class Rule:
+    is_temp = None
+    version = None
+    old_url = None
+    new_url = None
+
+def parse_versions(defs):
+    for x in range (0, len(defs)):
+        if defs[x][0] == 'versions':
+            return defs[x][1]
+
+def generate_rule(is_temp, version, old_url, new_url, is_symlink = False):
+
+    global definitions
+    global symlinks
+    global rules
+    
+    context_url = ''
+    
+    for x in range (0, len(definitions)):
+        if (definitions[x][0] == 'base'):
+            context_url = definitions[x][1]
+    
+    # if url contains {version} - substitute in the correct version
+    old_url_sub = rule_substitue(old_url, version)
+    new_url_sub = rule_substitue(new_url, version)
+    
+    # small reformatting the old url
+    if len(old_url_sub) > 0:
+        if old_url_sub[0] != '/':
+            old_url_sub = '/' + old_url_sub
+            
+        if not is_symlink:
+            if old_url_sub[-1] == '/':
+                old_url_sub = old_url_sub[:-1]
+            old_url_sub = '/' + context_url + old_url_sub
+
+    new_rule = Rule()
+    
+    new_rule.is_temp = is_temp
+    new_rule.version = version
+    new_rule.old_url = old_url_sub
+    new_rule.new_url = new_url_sub
+    
+    # check for symlinks
+    if len(symlinks) > 0:
+        for x in range (len(symlinks)):
+            if version == symlinks[x][1]:
+                r = generate_rule(is_temp, symlinks[x][0], old_url, new_url, True)
+                rules.append(r)
+    
+    return new_rule
+
+def rule_substitue(url_string, version):
+    
+    global definitions
```

Comment:
Nix these globals. I'd recommend making a class called something like `RedirectContext` that contains these instance attributes, and anything operating on the context would become a method of that class.